Add optional queue parameter to Reactive providers

haritowa opened the pull request:
Hello, i implemented callback queue propagation for reactive providers(which address https://github.com/Moya/Moya/issues/1032 https://github.com/Moya/Moya/issues/1048 https://github.com/Moya/Moya/pull/762). I have same complains, that should be resolved before merge.

## Queue vs Scheduler
I saw several suggestions to use Schedulers instead of queue for RxSwift(and whatever abstraction ReactiveSwift has), but I’m still believe this queue is not what rx provider responsible for, this is Alamofire threading stuff, users may be confused why they should pass scheduler instead of subscribe/observe on(and maybe this moment should be reflected at method comment).

  We can’t use `SchedulerType`, case we need DispatchQueue, which unnecessary for abstract scheduler, so we have to constraint which scheduler type user can pass(so, instead of reuse his scheduler, user may be forced to create new just to pass queue).

RxSwift do several optimizations for Schedulers(and i’m believe ReactiveSwift too), for example it “converts” Concurrent queue to serial(creating new serial queue with old queue as target), which queue should we use in this case(and even, can we still access concurrent queue)? Not sure we can rely on internal queue in future "optimizations" 😄 

Everything previously said was my own thoughts, but the fact Schedulers configuration is internal(which means we can’t access it outside of module and 3 paragraphs above has no sense) indicates that Krunoslav Zaher don’t want us to use internal queue :)

## Comments
Maybe we should add reacher comments to method, explicitly indicating that we nor observe, nor subscribe on this queue, we just pass it to Alamofire as a callback queue. If user really want subscription(request method call) to run on queue - he should subscribe on caller side.  User may create provider with queue and forget about it, which leads to unexpected element generation queue. In current implementation there is no way to pass nil queue to Alamofire, if provider has default queue, and I don’t know is it okay.

## Other stuff
As far as I understand, I should update Readme(as at [this](https://github.com/Moya/Moya/pull/762) PR). Maybe I shouldn't use `??` operator? Seems it can affect compilation time, so maybe we should sacrifice readability in favour of compilation speed?

haritowa commented:
I'll fix lint issues later, when there will be no other reason to reject PR(case code can change during PR :))

AndrewSB commented on Sources/RxMoya/RxMoyaProvider.swift:
```diff
@@ -6,22 +6,35 @@ import Moya
+    
     /// Initializes a reactive provider.
-    override public init(endpointClosure: @escaping EndpointClosure = MoyaProvider.defaultEndpointMapping,
+    public init(endpointClosure: @escaping EndpointClosure = MoyaProvider.defaultEndpointMapping,
```
why is override no longer needed here? is it because you added the `queue` param? If so, should we not also add the `queue` param to the regular callback `MoyaProvider`?

haritowa replied:
Maybe, I'm not sure 'case by some reason, queue wasn't in init. 
By the way, maybe we should create second init(return previous init, w\o queue and deprecate it? If anyone has reactive providers subclasses with overridden init - their code would not compile, sounds bad)

AndrewSB replied:
I don't think we'd need to deprecate the old signature, if we have a default value,`nil`, for the `queue` param, then it would be a non-breaking change. So it should just be safe to change the `MoyaProvider` initializer 😄

haritowa replied:
Sorry, i'm afraid you didn't get me. You are right, old init calls will not break after edit(case queue param is optional), but any RxSwift/ReactiveSwift provider descendants overridden init will not compile, until user remove `override`. Let me show you what i'm talking about:
#### This is Moya in current state with two user created provider descendants
![This is Moya in current Moya state with two user created provider descendants](https://cloud.githubusercontent.com/assets/6107211/25172990/53224cee-24fb-11e7-8d94-a4a84d460ed9.png)

#### This is Moya state, after adding queue param(note that new param has default value)
![screen shot 2017-04-19 at 12 22 07 pm](https://cloud.githubusercontent.com/assets/6107211/25173091/95c98fc6-24fb-11e7-88fa-1fddbf5f7c4e.png)

#### And finally that is was i proposed in previous commit
![screen shot 2017-04-19 at 12 22 52 pm](https://cloud.githubusercontent.com/assets/6107211/25173106/a4e831ba-24fb-11e7-905d-b970e9ccd20b.png)

Adding queue param to `MoyaProvider` will do even more mess, if we will stick to replace instead of deprecation

AndrewSB replied:
Oh!! I see, interesting behavior. Thank you for explaining it with examples! I totally get it now

Let's just make a breaking change and target 9.0.0 instead? What are your thoughts? Do you think deprecating is better practice?

haritowa replied:
If we can choose between shift breaking change to major upgrade or generate boilerplate with deprecating init - i prefer first approach :)

AndrewSB replied:
Yes, I think adding it to all 3 is the best approach, so we have feature parity (as much as possible) between all three.
We also might be able to remove all the `queue` logic from our Reactive and Rx subclasses by instead implementing it in MoyaProvider and then doing super.request instead of self.request in places like https://github.com/Moya/Moya/pull/1055/files#diff-ffd4342ff685cb2dff7f9c3cf77cef22R38

Let's do the deprecate approach, and then maybe in 9.0.0 I can remove the deprecated function so we keep our codebase clean (I can make that PR after we're done with this one)

AndrewSB commented:
Let's get the linting issues fixed, maybe write some tests. 

Then we can merge this into `optional-response-thread-param`, and continue the discussion on which queue/scheduler primitive to use in #762

haritowa commented:
> I think it would be smart to implement some more tests as well though. Are you familiar with testing in ReactiveSwift/RxSwift?

Yep, for some degree, but only with RxTest/RxBlocking(that means I can't write ReactiveSwift tests w/o future investigation). I'll cover this commit functionality(check callback queue) within the next few days

AndrewSB commented:
Cool, it would be awesome to get those RxSwift tests then! Maybe someone else can add them for ReactiveSwift later.

Things pending before we merge to `optional-response-thread-param`
- [ ] write tests for RxSwift
- [ ] fix the overriding initializer and change the `MoyaProvider` initializer to take a queue as well

justinmakaila commented:
So I'm trying to understand... this is the queue that will be used by Alamofire for callbacks as a hook back into Moya, whereas the remaining Scheduler property is in place for delivery of whatever reactive abstraction you use?

AndrewSB commented:
That's correct @justinmakaila. So then if you're using the vanilla `MoyaProvider`, you pass a thread. But if you're using an `RxMoyaProvider`, you'd pass in the Rx abstraction of a thread - a scheduler, and whatever the equivalent is for ReactiveSwift as well.

For rationale and example usage, check out #762

haritowa commented:
As far as I remember, you can't get queue from a scheduler. In addition, you can't pass any Scheduler as a callback queue provider for Alamofire

AndrewSB commented:
Let's re-examine that assumption: that you can't get a queue from a scheduler.
Schedulers internally use DispatchQueues to do their work ([here's](https://github.com/ReactiveX/RxSwift/blob/master/RxSwift/Schedulers/MainScheduler.swift#L23) how the MainScheduler uses a queue).

Let's make an issue on RxSwift's repo asking if theres any way to grab it, since right now, it is private

haritowa commented:
It's not private, it's internal. You can implement scheduler in whatever way you want: you should just adopt one protocol, which has nothing common with GCD. So, for instance, we need scheduler which stick with one thread, we can't get it's queue 'case it doesn't use any. There is sync and async versions for main scheduler(and again, we are talking about Schedulers, provided out of the box), but Alamofire wouldn't respect this distinction.
As I said earlier: we don't use this queue for any Rx-related work, we just pass to non-rx world. We don't even guarantee, that each element will be emitted on this scheduler(because we has dependency on Alamofire implementation). If user really want to use scheduler - he can just `observeOn` it any rx-chain he want, but he can't avoid dispatching to main queue for now. 
Anyway, I remember about tests and moving queue to MoyaProvider :)

justinmakaila commented:
Yeah I agree. The queue passed to Alamofire doesn't necessarily have to be the same one as our Reactive `Scheduler`, all that really matters is that we provide this init option so people can move their request building/handling off of the main queue and into a dedicated background queue (if desired).

haritowa commented:
Hmm, any clue why tests failed? I can't even open log. By the way, i added tests for `MoyaProvider` `RxMoyaProvider` and moved queue to `MoyaProvider`. Deprecate approach doesn't work as i expected(swift can't choose proper overload)

AndrewSB commented:
@haritowa: yeah, CI failed because there's no changelog for this. We're working on making that more clear to contributors.

For now, can you just add an entry to the changelog, and then we can see how the tests do?

haritowa commented:
@AndrewSB Oh i see, i was afraid you just give up on this PR :) Ok, i'll update changelog

AndrewSB commented:
I got really busy with work, sorry about that @haritowa!

haritowa commented:
@AndrewSB It's ok, I'm busy on weekdays too :)

haritowa commented:
@AndrewSB Aaaaand... it's done :)

sunshinejr commented:
I'm wondering if the parameter name `queue` isn't too vague and e.g. `callbackQueue` wouldn't help with recognizing for what exactly is this queue used. Thoughts?

haritowa commented:
@sunshinejr I took this name from original request method, but I'm agree, callbackQueue would be much better(I believe "queue" parameter name misleading even in request method)

sunshinejr commented:
Cool! What do you guys think @AndrewSB, @justinmakaila?

BasThomas commented:
Yes, I think `callbackQueue` is a better name. It takes away any ambiguity.

@haritowa, do you think we can improve the `queue` parameter in the `request` method as well?

haritowa commented:
@BasThomas Yep, I think so. In method context `queue` is less confusing, but still does not provide any clue about its mission

BasThomas commented:
What do you think is a better option?

haritowa commented:
I prefer slightly more verbose `callbackQueue`

pedrovereza commented:
The CI failure is legit, seems like `RxSwiftMoyaProviderTests`needs a second review 😄

AndrewSB commented:
that was broken by me -- I'll fix it now. Sorry!

sunshinejr commented:
Seems like @AndrewSB fixed the build! Things to do before the merge: 
- Fix the **Chanelog.md** conflicts (basically add the new line to **Next** column instead of **8.0.4**, because **8.0.4** was released)
- Change the name of `queue` to `callbackQueue` (wherever we can)

Does that seem right? 😄

AndrewSB commented:
yup, that does! Then we can figure out docs and usage instructions in #762

sunshinejr commented:
@haritowa Are these two points, I wrote above, something you have time to do in the near future? If not, I might try to help :)

pedrovereza commented:
@sunshinejr @AndrewSB @haritowa I fixed the conflict in the changelog (used Github web interface to do it, that's why the commit message is bad 😕 )

Does it make sense to merge this PR into `optional-response-thread-param` and take the `queue` renaming (and docs) in #762? 🤔

sunshinejr commented:
I think so 👍 Probably in another PR tho, that would point to `optional-response-thread-param`. But I'm in.

AndrewSB commented:
Yeah, lets merge this into `optional-response-thread-param`, and then do another PR also targeted to `optional-response-thread-param` that handles the renaming and docs 👌

ashfurrow commented:
Thanks a lot for contributing to Moya! I've invited you to join the Moya GitHub organization – no pressure to accept! If you'd like more information on what that means, check out our [contributor guidelines](https://github.com/Moya/contributors) and feel free to reach out with any questions.
